CovertInfiltration/Src/CovertInfiltration/Classes/UIListener_CovertActionCompleted.uc

```diff
@@ -0,0 +1,55 @@
+//---------------------------------------------------------------------------------------
+//  AUTHOR:  statusNone
+//  PURPOSE: This class is used to track when a covert action is completed in order to
+//	apply soldier will loss
+//---------------------------------------------------------------------------------------
+//  WOTCStrategyOverhaul Team
+//---------------------------------------------------------------------------------------
+
+class UIListener_CovertActionCompleted extends UIScreenListener config(Infiltration);
+
+var config float MIN_WILL_LOSS;
+var config float MAX_WILL_LOSS;
+
+event OnInit(UIScreen Screen)
+{
+	local XComGameState_CovertAction CovertAction;
+	local StateObjectReference Action;
+
+	if (Screen != UICovertActionReport(Screen)) return;
+	
+	Action = UICovertActionReport(Screen).ActionRef;
+	CovertAction = XComGameState_CovertAction(`XCOMHISTORY.GetGameStateForObjectID(Action.ObjectID));
+
+	ApplyWillLossToSoldiers(CovertAction);
+}
+
+function ApplyWillLossToSoldiers(XComGameState_CovertAction CovertAction)
+{
+	local XComGameState NewGameState;
+	local XComGameState_StaffSlot SlotState;
+	local XComGameState_Unit UnitState;
+
+	local int idx;
+
+	NewGameState = class'XComGameStateContext_ChangeContainer'.static.CreateChangeState("CI: Applying will loss to soldiers");
+
+	for (idx = 0; idx < CovertAction.StaffSlots.Length; idx++)
+	{
+		SlotState = CovertAction.GetStaffSlot(idx);
+		UnitState = SlotState.GetAssignedStaff();
+
+		if (SlotState.IsSlotFilled() && UnitState.IsSoldier())
+		{
+			UnitState.SetCurrentStat(eStat_Will, GetWillLoss(UnitState.GetMaxStat(eStat_Will)));
+			UnitState.UpdateMentalState();
```

Comment:
This ends up writing to the History. You need to use `NewGameState.ModifyStateObject` to properly modify units.

Additionally, you should make sure that the unit can actually recover will by adding / updating a `XComGameState_HeadquartersProjectRecoverWill`, like [here](https://github.com/X2CommunityCore/X2WOTCCommunityHighlander/blob/0c6933ce423d226ad94a1584b6a84cb9d80e5531/X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameStateContext_StrategyGameRule.uc#L1004-L1019):

			// First remove existing recover will project if there is one.
			XComHQ = GetAndAddXComHQ(NewGameState);
			foreach History.IterateByClassType(class'XComGameState_HeadquartersProjectRecoverWill', WillProject)
			{
				if(WillProject.ProjectFocus == UnitState.GetReference())
				{
					XComHQ.Projects.RemoveItem(WillProject.GetReference());
					NewGameState.RemoveStateObject(WillProject.ObjectID);
					break;
				}
			}


			// Add new will recover project
			WillProject = XComGameState_HeadquartersProjectRecoverWill(NewGameState.CreateNewStateObject(class'XComGameState_HeadquartersProjectRecoverWill'));
			WillProject.SetProjectFocus(UnitState.GetReference(), NewGameState);
			XComHQ.Projects.AddItem(WillProject.GetReference());